39854: Fix city name validation to allow digits, &, ., () characters

bhushan-cs opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Fixed city name validation to allow commonly used characters that were being incorrectly rejected. The current validation was too restrictive and prevented valid city names containing numbers, periods, ampersands, and parentheses.


### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#39854

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Try to place an order with city names like:
   - "St. Petersburg"
   - "Trinidad & Tobago"
   - "Paris (75001)"
2. Verify checkout completes successfully

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

bhushan-cs committed:
39854: Fix city name validation to allow digits, &, ., () characters

gwharton commented:
You might want to update the error message as well so users are informed of the correct characters that can be used.

bhushan-cs commented:
@gwharton thanks for the feedback! That's a great point. I think the following will be good. What do you think?
`Invalid City. Please use letters, numbers, spaces, and the following characters: - ' . , & ( )`

gwharton commented:
Looks good to me. I may expect some push back from Magento team on this, as the explicitly used to allow commas, then they removed commas as allowed in 2.4.8-beta1, so I'm figuring there must have been a reason why they tightened this down. They'll have to advise.

I would be willing to concede that comma does not belong in the city name, as comma is commonly used as a separator for address fields, but I can't see a reason why the others shouldn't be allowed, especially ampersand and full stop which should 100% be allowed IMHO.

hostep commented:
@gwharton, it wasn't Magento core devs that added this check, it was done by the community in https://github.com/magento/magento2/pull/38345

So maybe @dekiakbar can take a look here and leave some feedback?

gwharton commented:
Looks good to me. I may expect some push back from Magento team on this, as the explicitly used to allow commas, then they removed commas as allowed in 2.4.8-beta1, so I'm figuring there must have been a reason why they tightened this down. They'll have to advise.

I would be willing to concede that comma does not belong in the city name, as comma is commonly used as a separator for address fields, but I can't see a reason why the others shouldn't be allowed, especially ampersand and full stop which should 100% be allowed IMHO.

Yep, looks like a discussion around security is needed again. There must be a better way of detecting someone entering code snippets into the address fields without banning full stop, comma, ampersand, parenthesis and digits.

None of my customers who live in "Brighton & Hove" city can place orders with me without misspelling their city name (using and instead of &). Same applies for my customers in "St. Davids".

bhushan-cs committed:
39854: Update city validation error message to match allowed characters

bhushan-cs commented:
Thanks for digging into the history, @hostep and @gwharton 
I agree that a solution that allows common characters like '&' and '.' while preventing malicious input would be ideal.
I'm interested to hear @dekiakbar & @Franciscof-Serfe's thoughts on this.

dekiakbar commented:
Hi All,

@hostep Yes, its from that change. I agree, my changes was too strict and we should not to make it strict like this

@bhushan-cs Your changes make this validation better and also the Error message more clear and easy to unserstang. I fully agree with you

@gwharton You're absolutely right, and I’m really sorry for the inconvenience this has caused. I completely agree — blocking common characters like `&`, `.`, `()` and `digits` in address fields isn’t practical, especially when they’re part of legitimate place names like "Brighton & Hove" or "St. Davids."

gwharton commented:
For those stores that have individual and local issues, I guess there's nothing stopping them overriding the existing city validator in the XML config and having a stricter version local to their store.

dekiakbar commented:
That's a great point, allowing stores to override the city validator locally makes a lot of sense. It gives flexibility for specific needs without affecting the global behavior

bhushan-cs committed:
39854: Add support for typographical apostrophe and underscore in city validation and update test cases

bhushan-cs commented:
Hi @Franciscof-Serfe and @gwharton,

Thanks for the further feedback and suggestions!

@Franciscof-Serfe, I agree that adding support for underscore (_) and the typographical apostrophe (’) would definitely improve flexibility and handle more real-world city names accurately. I've updated the regular expression to include these.

@gwharton, I've also added the appropriate test cases to ensure these new additions are covered.

bhushan-cs committed:
Merge branch '2.4-develop' into 39854-fix-city-name-validation

hostep commented:
Hi guys

I ran the old and new regex on top of some databases of our bigger clients, to see if all cases are fixed now.

Results:
- for city field, many issues with the old regex, but no issues found with the latest regex presented in this PR, nice!
- for street fields, no issues found
- for telephone field:
  - I see a lot of customers using `/` or `.` symbols, I would allow those, they are very often used, example: `050/11.22.33`
  - I also found some telephone numbers that are longer then 20 characters, so maybe we should increase it to max 25 characters? Example: `+ 33 (0)6 11 22 33 44`

Proposal for new regex for telephone field: `private const PATTERN_TELEPHONE = '/(?:[\d\s\+\-\.\/\()]{1,25})/u';`
Not sure if we should include a better regex for the telephone field in this PR, or if we should open a new one?

gwharton commented:
To add to the telephone conversation, telephone numbers provided by Amazon USA marketplace are of the format 

"+1 111-111-1111 ext. 11111"

which is 26 characters. I remove the full stop to get them down to 25 characters and enter them manually in admin to bypass the telephone check at the moment which means they are compliant with Royal Mail restrictions which is 25 max.

engcom-Bravo commented:
Hi @bhushan-cs,

Thank you for your contribution! The Magento core engineering team is working on the issue which you have addressed in this PR. Team will cherry pick the commits from your PR and may do further implementation to cover few more scenarios as needed. We will reach out to you if we need more information. For now, you can pause work on this PR.

 Thank you once again!
